File: Classes/TiGooglemapsView.m
/**
 * Ti.GoogleMaps
 * Copyright (c) 2015-present by Hans Knöchel. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsView.h"
#import "TiClusterIconGenerator.h"
#import "TiGooglemapsAnnotationProxy.h"
#import "TiGooglemapsCircleProxy.h"
#import "TiGooglemapsClusterItemProxy.h"
#import "TiGooglemapsConstants.h"
#import "TiGooglemapsPolygonProxy.h"
#import "TiGooglemapsPolylineProxy.h"
#import "TiGooglemapsViewProxy.h"
#import "TiPOIItem.h"

static NSString *const kGMUMyLocationKeyPath = @"myLocation";

@implementation TiGooglemapsView

- (void)dealloc
{
  [_mapView removeObserver:self forKeyPath:kGMUMyLocationKeyPath];
  _mapView.delegate = nil;
  _clusterRenderer.delegate = nil;
}

- (GMSMapView *)mapView
{
  if (_mapView == nil) {
    _mapView = [[GMSMapView alloc] initWithFrame:[self bounds]];
    [_mapView setMapType:kGMSTypeNormal];
    [_mapView setDelegate:self];
    [_mapView setAutoresizingMask:UIViewAutoresizingNone];

    [_mapView addObserver:self
               forKeyPath:kGMUMyLocationKeyPath
                  options:NSKeyValueObservingOptionNew
                  context:nil];

    [self addSubview:_mapView];
  }

  return _mapView;
}

- (GMUClusterManager *)clusterManager
{
  if (_clusterManager == nil) {
    // Set up the cluster manager with default icon generator and renderer.
    id<GMUClusterAlgorithm> algorithm = [[GMUNonHierarchicalDistanceBasedAlgorithm alloc] init];

    TiClusterIconGenerator *iconGenerator = [self createIconGenerator];

    _clusterRenderer = [[TiClusterRenderer alloc] initWithMapView:_mapView clusterIconGenerator:iconGenerator];
    _clusterRenderer.delegate = self;

    _clusterManager = [[GMUClusterManager alloc] initWithMap:[self mapView] algorithm:algorithm renderer:_clusterRenderer];
    [_clusterManager setDelegate:self mapDelegate:self];
  }

  return _clusterManager;
}

- (void)renderer:(id<GMUClusterRenderer>)renderer willRenderMarker:(GMSMarker *)marker
{
  if ([[marker userData] isKindOfClass:[TiPOIItem class]]) {
    TiPOIItem *item = (TiPOIItem *)[marker userData];

    // Note: All native props are nullable, so we don't need to check against nil here

    // TODO: Add more property mappings here
    [marker setTitle:item.title];
    [marker setSnippet:item.subtitle];
    [marker setIcon:item.icon];
  }
}

- (TiClusterIconGenerator *)createIconGenerator
{
  id clusterRanges = [[self proxy] valueForKey:@"clusterRanges"];
  id clusterBackgrounds = [[self proxy] valueForKey:@"clusterBackgrounds"];

  if (clusterRanges && clusterBackgrounds) {
    NSMutableArray *backgrounds = [NSMutableArray array];

    for (id background in clusterBackgrounds) {
      ENSURE_TYPE(background, NSString);
      UIImage *clusterBackground = [TiUtils image:background proxy:self.proxy];

      if (!clusterBackground) {
        NSLog(@"[ERROR] Cluster background-file (%@) cannot be found, skipping ...");
        continue;
      }

      [backgrounds addObject:clusterBackground];
    }

    return [[TiClusterIconGenerator alloc] initWithBuckets:clusterRanges backgroundImages:backgrounds];
  } else if (clusterRanges) {
    return [[TiClusterIconGenerator alloc] initWithBuckets:clusterRanges];
  }

  return [[TiClusterIconGenerator alloc] init];
}

- (TiGooglemapsViewProxy *)mapViewProxy
{
  return (TiGooglemapsViewProxy *)[self proxy];
}

- (void)frameSizeChanged:(CGRect)frame bounds:(CGRect)bounds
{
  [TiUtils setView:_mapView positionRect:bounds];
  [super frameSizeChanged:frame bounds:bounds];
}

#pragma mark Cluster Delegates

- (BOOL)clusterManager:(GMUClusterManager *)clusterManager didTapCluster:(id<GMUCluster>)cluster
{
  if ([[self proxy] _hasListeners:@"clusterclick"]) {
    [[self proxy] fireEvent:@"clusterclick"
                 withObject:@{
                   @"latitude" : @(cluster.position.latitude),
                   @"longitude" : @(cluster.position.longitude),
                   @"count" : @(cluster.count),
                   @"clusterItems" : [self arrayFromClusterItems:cluster.items]
                 }];
  }

  GMSCameraPosition *newCamera = [GMSCameraPosition cameraWithTarget:cluster.position zoom:_mapView.camera.zoom + 1];
  [_mapView moveCamera:[GMSCameraUpdate setCamera:newCamera]];

  return YES;
}

- (BOOL)clusterManager:(GMUClusterManager *)clusterManager didTapClusterItem:(id<GMUClusterItem>)clusterItem
{
  if ([[self proxy] _hasListeners:@"clusteritemclick"]) {
    [[self proxy] fireEvent:@"clusteritemclick"
                 withObject:@{
                   @"latitude" : @(clusterItem.position.latitude),
                   @"longitude" : @(clusterItem.position.longitude),
                   @"title" : [(TiPOIItem *)clusterItem title] ?: [NSNull null],
                   @"subtitle" : [(TiPOIItem *)clusterItem subtitle] ?: [NSNull null],
                   @"userData" : [(TiPOIItem *)clusterItem userData] ?: [NSNull null]
                 }];
  }

  return YES;
}

#pragma mark Map View Delegates

- (void)mapView:(GMSMapView *)mapView willMove:(BOOL)gesture
{
  if ([[self proxy] _hasListeners:@"regionwillchange"]) {
    [[self proxy] fireEvent:@"regionwillchange"
                 withObject:@{
                   @"map" : [self proxy],
                   @"latitude" : @(mapView.camera.target.latitude),
                   @"longitude" : @(mapView.camera.target.longitude),
                   @"gesture" : @(gesture)
                 }];
  }
}

- (void)mapView:(GMSMapView *)mapView didChangeCameraPosition:(GMSCameraPosition *)position
{
  if ([[self proxy] _hasListeners:@"regionchanged"]) {
    NSMutableDictionary *updatedRegion = [NSMutableDictionary dictionaryWithDictionary:@{
      @"latitude" : @(position.target.latitude),
      @"longitude" : @(position.target.longitude),
      @"zoom" : @(position.zoom),
      @"bearing" : @(position.bearing),
      @"viewingAngle" : @(position.viewingAngle)
    }];

    [(TiGooglemapsViewProxy *)[self proxy] replaceValue:updatedRegion forKey:@"region" notification:NO];

    NSMutableDictionary *event = [NSMutableDictionary dictionaryWithDictionary:updatedRegion];
    [event setObject:[self proxy] forKey:@"map"];

    [[self proxy] fireEvent:@"regionchanged" withObject:event];
  }
}

- (void)mapView:(GMSMapView *)mapView idleAtCameraPosition:(GMSCameraPosition *)position
{
  if ([[self proxy] _hasListeners:@"idle"]) {
    [[self proxy] fireEvent:@"idle" withObject:[self dictionaryFromCameraPosition:position]];
  }
}

- (void)mapView:(GMSMapView *)mapView didTapAtCoordinate:(CLLocationCoordinate2D)coordinate
{
  if ([[self proxy] _hasListeners:@"mapclick"]) {
    [[self proxy] fireEvent:@"mapclick"
                 withObject:@{
                   @"map" : [self proxy],
                   @"latitude" : @(coordinate.latitude),
                   @"longitude" : @(coordinate.longitude)
                 }];
  }
}

- (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordinate2D)coordinate
{
  if ([[self proxy] _hasListeners:@"longclick"]) {
    [[self proxy] fireEvent:@"longclick"
                 withObject:@{
                   @"map" : [self proxy],
                   @"latitude" : @(coordinate.latitude),
                   @"longitude" : @(coordinate.longitude)
                 }];
  }
}

- (BOOL)mapView:(GMSMapView *)mapView didTapMarker:(GMSMarker *)marker
{
  if ([[self proxy] _hasListeners:@"click"]) {
    [[self proxy] fireEvent:@"click"
                 withObject:@{
                   @"clicksource" : @"pin",
                   @"annotation" : [self dictionaryFromMarker:marker],
                   @"map" : [self proxy],
                   @"latitude" : @(marker.position.latitude),
                   @"longitude" : @(marker.position.longitude)
                 }];
  }

  return NO;
}

- (void)mapView:(GMSMapView *)mapView didTapInfoWindowOfMarker:(GMSMarker *)marker
{
  if ([[self proxy] _hasListeners:@"click"]) {
    [[self proxy] fireEvent:@"click"
                 withObject:@{
                   @"clicksource" : @"infoWindow",
                   @"annotation" : [self dictionaryFromMarker:marker],
                   @"map" : [self proxy],
                   @"latitude" : @(marker.position.latitude),
                   @"longitude" : @(marker.position.longitude)
                 }];
  }
}

- (void)mapView:(GMSMapView *)mapView didTapOverlay:(GMSOverlay *)overlay
{
  if ([[self proxy] _hasListeners:@"overlayclick"]) {
    [[self proxy] fireEvent:@"overlayclick"
                 withObject:@{
                   @"overlay" : [self overlayProxyFromOverlay:overlay]
                 }];
  }
  if ([[self proxy] _hasListeners:@"click"]) {
    [[self proxy] fireEvent:@"click"
                 withObject:@{
                   @"clicksource" : [self overlayTypeFromOverlay:overlay],
                   @"map" : [self proxy],
                   @"overlay" : [self overlayProxyFromOverlay:overlay]
                 }];
  }
}

- (void)mapView:(GMSMapView *)mapView didTapPOIWithPlaceID:(NSString *)placeID name:(NSString *)name location:(CLLocationCoordinate2D)location
{
  if ([[self proxy] _hasListeners:@"poiclick"]) {
    NSDictionary *event = @{
      @"placeID" : placeID,
      @"name": name,
      @"latitude" : @(location.latitude),
      @"longitude" : @(location.longitude)
    };
    [[self proxy] fireEvent:@"poiclick" withObject:event];
  }
}

- (void)mapView:(GMSMapView *)mapView didBeginDraggingMarker:(GMSMarker *)marker
{
  if ([[self proxy] _hasListeners:@"dragstart"]) {
    NSDictionary *event = @{
      @"annotation" : [self dictionaryFromMarker:marker]
    };
    [[self proxy] fireEvent:@"dragstart" withObject:event];
  }
}

- (void)mapView:(GMSMapView *)mapView didEndDraggingMarker:(GMSMarker *)marker
{
  if ([[self proxy] _hasListeners:@"dragend"]) {
    NSDictionary *event = @{
      @"annotation" : [self dictionaryFromMarker:marker]
    };
    [[self proxy] fireEvent:@"dragend" withObject:event];
  }
}

- (void)mapView:(GMSMapView *)mapView didDragMarker:(GMSMarker *)marker
{
  if ([[self proxy] _hasListeners:@"dragmove"]) {
    NSDictionary *event = @{
      @"annotation" : [self dictionaryFromMarker:marker]
    };
    [[self proxy] fireEvent:@"dragmove" withObject:event];
  }
}

- (UIView *)mapView:(GMSMapView *)mapView markerInfoWindow:(GMSMarker *)marker
{
  TiGooglemapsAnnotationProxy *annotation = nil;

  for (TiGooglemapsAnnotationProxy *proxy in [[(TiGooglemapsViewProxy *)[self proxy] markers] mutableCopy]) {
    if ([[[[proxy marker] userData] objectForKey:@"uuid"] isEqualToString:[[marker userData] objectForKey:@"uuid"]]) {
      annotation = proxy;
    }
  }

  if (!annotation || !annotation.infoWindow) {
    return nil;
  }
  
  [annotation rememberSelf];

  return [[annotation infoWindow] view];
}

- (BOOL)didTapMyLocationButtonForMapView:(GMSMapView *)mapView
{
  if ([[self proxy] _hasListeners:@"locationclick"]) {
    NSDictionary *event = @{
      @"map" : [self proxy]
    };
    [[self proxy] fireEvent:@"locationclick" withObject:event];
  }

  return NO;
}

- (void)mapViewDidFinishTileRendering:(GMSMapView *)mapView
{
  if ([[self proxy] _hasListeners:@"complete"]) {
    [[self proxy] fireEvent:@"complete"];
  }
}

#pragma mark Helper

- (NSDictionary *)dictionaryFromCameraPosition:(GMSCameraPosition *)position
{
  if (position == nil) {
    return @{};
  }

  return @{
    @"latitude" : @(position.target.latitude),
    @"longitude" : @(position.target.longitude),
    @"zoom" : @(position.zoom),
    @"viewingAngle" : @(position.viewingAngle),
    @"bearing" : @([position bearing])
  };
}

- (NSDictionary *)dictionaryFromCoordinate:(CLLocationCoordinate2D)coordinate
{
  return @{
    @"latitude" : @(coordinate.latitude),
    @"longitude" : @(coordinate.longitude)
  };
}

- (NSDictionary *)dictionaryFromMarker:(GMSMarker *)marker
{
  if (!marker) {
    return @{};
  }

  return @{
    @"latitude" : @(marker.position.latitude),
    @"longitude" : @(marker.position.longitude),
    @"userData" : marker.userData ?: [NSNull null],
    @"title" : marker.title ?: [NSNull null],
    @"subtitle" : marker.snippet ?: [NSNull null]
  };
}

- (id)overlayTypeFromOverlay:(GMSOverlay *)overlay
{
  ENSURE_UI_THREAD(overlayTypeFromOverlay, overlay);

  if ([overlay isKindOfClass:[GMSPolygon class]]) {
    return @(TiGooglemapsOverlayTypePolygon);
  } else if ([overlay isKindOfClass:[GMSPolyline class]]) {
    return @(TiGooglemapsOverlayTypePolyline);
  } else if ([overlay isKindOfClass:[GMSCircle class]]) {
    return @(TiGooglemapsOverlayTypeCircle);
  }

  NSLog(@"[ERROR] Unknown overlay provided: %@", [overlay class])

      return @(TiGooglemapsOverlayTypeUnknown);
}

- (id)overlayProxyFromOverlay:(GMSOverlay *)overlay
{
  for (TiProxy *overlayProxy in [[self mapViewProxy] overlays]) {
    // Check for polygons
    if ([overlay isKindOfClass:[GMSPolygon class]] && [overlayProxy isKindOfClass:[TiGooglemapsPolygonProxy class]]) {
      if ([(TiGooglemapsPolygonProxy *)overlayProxy polygon] == overlay) {
        return (TiGooglemapsPolygonProxy *)overlayProxy;
      }

      // Check for polylines
    } else if ([overlay isKindOfClass:[GMSPolyline class]] && [overlayProxy isKindOfClass:[TiGooglemapsPolylineProxy class]]) {
      if ([(TiGooglemapsPolylineProxy *)overlayProxy polyline] == overlay) {
        return (TiGooglemapsPolylineProxy *)overlayProxy;
      }

      // Check for circles
    } else if ([overlay isKindOfClass:[GMSCircle class]] && [overlayProxy isKindOfClass:[TiGooglemapsCircleProxy class]]) {
      if ([(TiGooglemapsCircleProxy *)overlayProxy circle] == overlay) {
        return (TiGooglemapsCircleProxy *)overlayProxy;
      }
    }
  }

  return [NSNull null];
}

- (NSArray *)arrayFromClusterItems:(NSArray<id<GMUClusterItem>> *)clusterItems
{
  NSMutableArray *result = [NSMutableArray arrayWithCapacity:clusterItems.count];

  for (id<GMUClusterItem> clusterItem in clusterItems) {
    [result addObject:[[TiGooglemapsClusterItemProxy alloc] _initWithPageContext:[[self proxy] pageContext]
                                                                     andPosition:clusterItem.position
                                                                           title:[(TiPOIItem *)clusterItem title]
                                                                        subtitle:[(TiPOIItem *)clusterItem subtitle]
                                                                            icon:[(TiPOIItem *)clusterItem icon]
                                                                        userData:[(TiPOIItem *)clusterItem userData]]];
  }

  return result;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context
{
  CLLocation *myLocation = _mapView.myLocation;

  if (myLocation != nil) {
    [[self proxy] fireEvent:@"myLocationUpdate" withObject:@{ @"latitude": @(myLocation.coordinate.latitude), @"longitude": @(myLocation.coordinate.longitude) }];
  }
}

#pragma mark Constants

MAKE_SYSTEM_PROP(OVERLAY_TYPE_POLYGON, TiGooglemapsOverlayTypePolygon);
MAKE_SYSTEM_PROP(OVERLAY_TYPE_POLYLINE, TiGooglemapsOverlayTypePolyline);
MAKE_SYSTEM_PROP(OVERLAY_TYPE_CIRCLE, TiGooglemapsOverlayTypeCircle);

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsView.swift
//
//  TiGooglemapsView.swift
//  titanium-googlemaps
//
//  Created by Your Name
//

import UIKit
import TitaniumKit
import GoogleMaps
import GoogleMapsUtils

private let kGMUMyLocationKeyPath = "myLocation"

@objc(TiGooglemapsView)
class TiGooglemapsView: TiUIView, GMSMapViewDelegate, GMUClusterManagerDelegate, GMUClusterRendererDelegate {
  private var centerAnnotationOnTap = true
  
  private var mapViewStorage: GMSMapView?
  private var clusterManagerStorage: GMUClusterManager?
  private var clusterRendererStorage: TiClusterRenderer?
  
  @objc var mapView: GMSMapView {
    if let mapView = mapViewStorage {
      return mapView
    }
    
    let options = GMSMapViewOptions()
    options.frame = frame
    options.backgroundColor = .systemBackground
    
    let mapView = GMSMapView(options: options)
    mapView.mapType = GMSMapViewType.normal
    mapView.delegate = self
    mapView.autoresizingMask = []
    mapView.addObserver(self, forKeyPath: kGMUMyLocationKeyPath, options: .new, context: nil)
    
    addSubview(mapView)
    mapViewStorage = mapView
    centerAnnotationOnTap = true

    return mapView
  }
  
  @objc var clusterRenderer: TiClusterRenderer {
    if let renderer = clusterRendererStorage {
      return renderer
    }
    
    let iconGenerator = createIconGenerator()
    let renderer = TiClusterRenderer(mapView: mapView, clusterIconGenerator: iconGenerator)
    renderer.delegate = self
    clusterRendererStorage = renderer
    return renderer
  }
  
  @objc var clusterManager: GMUClusterManager {
    if let clusterManager = clusterManagerStorage {
      return clusterManager
    }
    
    let algorithm = GMUNonHierarchicalDistanceBasedAlgorithm()
    let manager = GMUClusterManager(map: mapView, algorithm: algorithm, renderer: clusterRenderer)
    manager.setDelegate(self, mapDelegate: self)
    clusterManagerStorage = manager
    return manager
  }
  
  deinit {
    if let mapView = mapViewStorage {
      mapView.removeObserver(self, forKeyPath: kGMUMyLocationKeyPath)
      mapView.delegate = nil
    }
    
    clusterRendererStorage?.delegate = nil
  }
  
  private func createIconGenerator() -> TiClusterIconGenerator {
    guard let proxy = proxy else { return TiClusterIconGenerator() }
    
    let clusterRanges = proxy.value(forKey: "clusterRanges") as? [NSNumber]
    let clusterBackgrounds = proxy.value(forKey: "clusterBackgrounds") as? [Any]
    
    if let ranges = clusterRanges, let backgrounds = clusterBackgrounds {
      var images: [UIImage] = []
      for background in backgrounds {
        guard let image = TiUtils.image(background, proxy: proxy) else {
          NSLog("[ERROR] Cluster background-file (%@) cannot be found, skipping ...", "\(background)")
          continue
        }
        
        images.append(image)
      }
      
      return TiClusterIconGenerator(buckets: ranges, backgroundImages: images)
    } else if let ranges = clusterRanges {
      return TiClusterIconGenerator(buckets: ranges)
    }
    
    return TiClusterIconGenerator()
  }
  
  private var mapViewProxy: TiGooglemapsViewProxy? {
    return proxy as? TiGooglemapsViewProxy
  }
  
  override func frameSizeChanged(_ frame: CGRect, bounds: CGRect) {
    if let mapView = mapViewStorage {
      TiUtils.setView(mapView, positionRect: bounds)
    }
    
    super.frameSizeChanged(frame, bounds: bounds)
  }
  
  // MARK: - Cluster Delegates
  
  func renderer(_ renderer: GMUClusterRenderer, willRenderMarker marker: GMSMarker) {
    guard let item = marker.userData as? TiPOIItem else { return }
    
    marker.title = item.title
    marker.snippet = item.subtitle
    marker.icon = item.icon
  }
  
  func clusterManager(_ clusterManager: GMUClusterManager, didTap cluster: GMUCluster) -> Bool {
    guard let proxy = proxy else { return false }
    
    if proxy._hasListeners("clusterclick") {
      proxy.fireEvent("clusterclick", with: [
        "latitude": cluster.position.latitude,
        "longitude": cluster.position.longitude,
        "count": cluster.count,
        "clusterItems": arrayFromClusterItems(cluster.items)
      ])
    }
    
    let newCamera = GMSCameraPosition.camera(withTarget: cluster.position, zoom: mapView.camera.zoom + 1)
    mapView.moveCamera(GMSCameraUpdate.setCamera(newCamera))
    
    return true
  }
  
  func clusterManager(_ clusterManager: GMUClusterManager, didTap clusterItem: GMUClusterItem) -> Bool {
    guard let proxy = proxy else { return false }
    guard let poiItem = clusterItem as? TiPOIItem else { return false }
    
    if proxy._hasListeners("clusteritemclick") {
      proxy.fireEvent("clusteritemclick", with: [
        "latitude": clusterItem.position.latitude,
        "longitude": clusterItem.position.longitude,
        "title": poiItem.title ?? NSNull(),
        "subtitle": poiItem.subtitle ?? NSNull(),
        "userData": poiItem.userData ?? NSNull()
      ] as [String: Any])
    }
    
    return true
  }
  
  // MARK: - Map View Delegates
  
  func mapView(_ mapView: GMSMapView, willMove gesture: Bool) {
    guard let proxy = proxy, proxy._hasListeners("regionwillchange") else { return }
    
    proxy.fireEvent("regionwillchange", with: [
      "map": proxy,
      "latitude": mapView.camera.target.latitude,
      "longitude": mapView.camera.target.longitude,
      "gesture": gesture
    ])
  }
  
  func mapView(_ mapView: GMSMapView, didChange position: GMSCameraPosition) {
    guard let proxy = proxy, proxy._hasListeners("regionchanged") else { return }
    
    var updatedRegion: [String: Any] = [
      "latitude": position.target.latitude,
      "longitude": position.target.longitude,
      "zoom": position.zoom,
      "bearing": position.bearing,
      "viewingAngle": position.viewingAngle
    ]
    
    (proxy as? TiGooglemapsViewProxy)?.replaceValue(updatedRegion, forKey: "region", notification: false)
    updatedRegion["map"] = proxy
    
    proxy.fireEvent("regionchanged", with: updatedRegion)
  }
  
  func mapView(_ mapView: GMSMapView, idleAt position: GMSCameraPosition) {
    guard let proxy = proxy, proxy._hasListeners("idle") else { return }
    proxy.fireEvent("idle", with: dictionaryFromCameraPosition(position))
  }
  
  func mapView(_ mapView: GMSMapView, didTapAt coordinate: CLLocationCoordinate2D) {
    guard let proxy = proxy, proxy._hasListeners("mapclick") else { return }
    proxy.fireEvent("mapclick", with: [
      "map": proxy,
      "latitude": coordinate.latitude,
      "longitude": coordinate.longitude
    ])
  }
  
  func mapView(_ mapView: GMSMapView, didLongPressAt coordinate: CLLocationCoordinate2D) {
    guard let proxy = proxy, proxy._hasListeners("longclick") else { return }
    proxy.fireEvent("longclick", with: [
      "map": proxy,
      "latitude": coordinate.latitude,
      "longitude": coordinate.longitude
    ])
  }
  
  func mapView(_ mapView: GMSMapView, didTap marker: GMSMarker) -> Bool {
    guard let proxy = proxy else { return centerAnnotationOnTap }
    
    if proxy._hasListeners("click") {
      proxy.fireEvent("click", with: [
        "clicksource": "pin",
        "annotation": dictionaryFromMarker(marker),
        "map": proxy,
        "latitude": marker.position.latitude,
        "longitude": marker.position.longitude
      ])
    }
    
    return centerAnnotationOnTap
  }
  
  func mapView(_ mapView: GMSMapView, didTapInfoWindowOf marker: GMSMarker) {
    guard let proxy = proxy, proxy._hasListeners("click") else { return }
    proxy.fireEvent("click", with: [
      "clicksource": "infoWindow",
      "annotation": dictionaryFromMarker(marker),
      "map": proxy,
      "latitude": marker.position.latitude,
      "longitude": marker.position.longitude
    ])
  }
  
  func mapView(_ mapView: GMSMapView, didTap overlay: GMSOverlay) {
    guard let proxy = proxy else { return }
    
    if proxy._hasListeners("overlayclick") {
      proxy.fireEvent("overlayclick", with: [
        "overlay": overlayProxyFromOverlay(overlay)
      ])
    }
    
    if proxy._hasListeners("click") {
      proxy.fireEvent("click", with: [
        "clicksource": overlayTypeFromOverlay(overlay),
        "map": proxy,
        "overlay": overlayProxyFromOverlay(overlay)
      ])
    }
  }
  
  func mapView(_ mapView: GMSMapView, didTapPOIWithPlaceID placeID: String, name: String, location: CLLocationCoordinate2D) {
    guard let proxy = proxy, proxy._hasListeners("poiclick") else { return }
    
    proxy.fireEvent("poiclick", with: [
      "placeID": placeID,
      "name": name,
      "latitude": location.latitude,
      "longitude": location.longitude
    ])
  }
  
  func mapView(_ mapView: GMSMapView, didBeginDragging marker: GMSMarker) {
    guard let proxy = proxy, proxy._hasListeners("dragstart") else { return }
    proxy.fireEvent("dragstart", with: ["annotation": dictionaryFromMarker(marker)])
  }
  
  func mapView(_ mapView: GMSMapView, didEndDragging marker: GMSMarker) {
    guard let proxy = proxy, proxy._hasListeners("dragend") else { return }
    proxy.fireEvent("dragend", with: ["annotation": dictionaryFromMarker(marker)])
  }
  
  func mapView(_ mapView: GMSMapView, didDrag marker: GMSMarker) {
    guard let proxy = proxy, proxy._hasListeners("dragmove") else { return }
    proxy.fireEvent("dragmove", with: ["annotation": dictionaryFromMarker(marker)])
  }
  
  func mapView(_ mapView: GMSMapView, markerInfoWindow marker: GMSMarker) -> UIView? {
    guard let annotationProxies = mapViewProxy?.markersCopy else { return nil }
    let markerUUID = (marker.userData as? [String: Any])?["uuid"] as? String
    
    let annotation = annotationProxies.first { proxy in
      guard let proxyUUID = (proxy.marker.userData as? [String: Any])?["uuid"] as? String else { return false }
      return proxyUUID == markerUUID
    }
    
    guard let infoAnnotation = annotation, let infoWindowProxy = infoAnnotation.infoWindow() else {
      return nil
    }
    
    infoAnnotation.rememberSelf()
    return infoWindowProxy.view
  }
  
  func didTapMyLocationButton(for mapView: GMSMapView) -> Bool {
    guard let proxy = proxy else { return false }
    
    if proxy._hasListeners("locationclick") {
      proxy.fireEvent("locationclick", with: ["map": proxy])
    }
    
    return false
  }
  
  func mapViewDidFinishTileRendering(_ mapView: GMSMapView) {
    guard let proxy = proxy else { return }
    
    if proxy._hasListeners("complete") {
      proxy.fireEvent("complete")
    }
    
    if proxy._hasListeners("ready") {
      proxy.fireEvent("ready")
    }
  }
  
  // MARK: - Helpers
  
  override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {
    guard keyPath == kGMUMyLocationKeyPath else {
      super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
      return
    }
    
    guard let location = mapView.myLocation else { return }
    proxy?.fireEvent("myLocationUpdate", with: [
      "latitude": location.coordinate.latitude,
      "longitude": location.coordinate.longitude
    ])
  }
  
  private func dictionaryFromCameraPosition(_ position: GMSCameraPosition?) -> [String: Any] {
    guard let position = position else { return [:] }
    
    return [
      "latitude": position.target.latitude,
      "longitude": position.target.longitude,
      "zoom": position.zoom,
      "viewingAngle": position.viewingAngle,
      "bearing": position.bearing
    ]
  }
  
  private func dictionaryFromMarker(_ marker: GMSMarker?) -> [String: Any] {
    guard let marker = marker else { return [:] }
    
    return [
      "latitude": marker.position.latitude,
      "longitude": marker.position.longitude,
      "userData": marker.userData ?? NSNull(),
      "title": marker.title ?? NSNull(),
      "subtitle": marker.snippet ?? NSNull()
    ]
  }
  
  private func overlayTypeFromOverlay(_ overlay: GMSOverlay?) -> NSNumber {
    guard let overlay = overlay else {
      return NSNumber(value: TiGooglemapsOverlayType.unknown.rawValue)
    }
    
    if overlay is GMSPolygon {
      return NSNumber(value: TiGooglemapsOverlayType.polygon.rawValue)
    } else if overlay is GMSPolyline {
      return NSNumber(value: TiGooglemapsOverlayType.polyline.rawValue)
    } else if overlay is GMSCircle {
      return NSNumber(value: TiGooglemapsOverlayType.circle.rawValue)
    }
    
    NSLog("[ERROR] Unknown overlay provided: %@", "\(type(of: overlay))")
    return NSNumber(value: TiGooglemapsOverlayType.unknown.rawValue)
  }
  
  private func overlayProxyFromOverlay(_ overlay: GMSOverlay) -> Any {
    guard let overlayProxies = mapViewProxy?.overlaysCopy else { return NSNull() }
    
    for proxy in overlayProxies {
      if let polygonProxy = proxy as? TiGooglemapsPolygonProxy,
         overlay === polygonProxy.polygon() {
        return polygonProxy
      } else if let polylineProxy = proxy as? TiGooglemapsPolylineProxy,
                overlay === polylineProxy.polyline() {
        return polylineProxy
      } else if let circleProxy = proxy as? TiGooglemapsCircleProxy,
                overlay === circleProxy.circle() {
        return circleProxy
      }
    }
    
    return NSNull()
  }
  
  private func arrayFromClusterItems(_ clusterItems: [GMUClusterItem]) -> [Any] {
    guard let context = proxy?.pageContext else { return [] }
    
    return clusterItems.compactMap { item in
      guard let poiItem = item as? TiPOIItem else { return nil }
     
      return TiGooglemapsClusterItemProxy()._init(
        withPageContext: context,
        andPosition: poiItem.position,
        title: poiItem.title,
        subtitle: poiItem.subtitle,
        icon: poiItem.icon,
        userData: poiItem.userData
      )
    }
  }
  
  @objc func setCenterAnnotationOnTap_(_ value: Any?) {
    centerAnnotationOnTap = TiUtils.boolValue(value)
  }
  
  @objc func OVERLAY_TYPE_POLYGON() -> NSNumber {
    return NSNumber(value: TiGooglemapsOverlayType.polygon.rawValue)
  }
  
  @objc func OVERLAY_TYPE_POLYLINE() -> NSNumber {
    return NSNumber(value: TiGooglemapsOverlayType.polyline.rawValue)
  }
  
  @objc func OVERLAY_TYPE_CIRCLE() -> NSNumber {
    return NSNumber(value: TiGooglemapsOverlayType.circle.rawValue)
  }
}
